Don't report an error when cleaning missing files

randycoulman opened the pull request:
When doing a `rake clean` or `rake clobber`, it is often the case that
some of the files don't exist.  The desired end state of the clean or
clobber task is that the specified files no longer be present at the
end of the task; if they weren't there in the first place, the
post-condition is still satisfied.

Without this change, running a `rake clean` or `rake clobber` is too
noisy, and yet silently failing to delete a file is not a good
solution either.

This may not be the best fix for this.  I tried two options:

1) Add a second rescue block for `Errno::ENOENT`.  This fails
`test_cleanup` in JRuby, because JRuby raises `Errno::ENOENT` for that
case.

2) Use `File.exist?` as I've done.  In order to make this work, I had
to modify the permissions used in `create_undeletable_file`, because
`File.exist?` returns false for a file in an unreadable directory.

With either of these solutions, there are corner cases where one or
more Ruby implementations will silently fail to delete a file.  My
opinion is that those corner cases are rare enough that this PR makes
sense for the far more common case.

randycoulman committed:
Don't report an error when cleaning missing files

When doing a `rake clean` or `rake clobber`, it is often the case that
some of the files don't exist.  The desired end state of the clean or
clobber task is that the specified files no longer be present at the
end of the task; if they weren't there in the first place, the
post-condition is still satisfied.

Without this change, running a `rake clean` or `rake clobber` is too
noisy, and yet silently failing to delete a file is not a good
solution either.

This may not be the best fix for this.  I tried two options:

1) Add a second rescue block for `Errno::ENOENT`.  This fails
`test_cleanup` in JRuby, because JRuby raises `Errno::ENOENT` for that
case.

2) Use `File.exist?` as I've done.  In order to make this work, I had
to modify the permissions used in `create_undeletable_file`, because
`File.exist?` returns false for a file in an unreadable directory.

With either of these solutions, there are corner cases where one or
more Ruby implementations will silently fail to delete a file.  My
opinion is that those corner cases are rare enough that this PR makes
sense for the far more common case.

drbrain commented:
I approve of this change in concept, but I dislike your solution.

I think in the rescue block rake should walk up the tree looking for a directory that exists but is not readable.  If one is found the message is displayed, otherwise nothing.

This way rake will not give a message about already deleted files.

randycoulman commented:
Sounds like a plan.  I'll work on that and resubmit.  I mostly just wanted to start the discussion, but figured it would be better to have some code to talk about.  Thanks!

randycoulman committed:
Be smarter about determining whether the file is already gone

Per @drbrain's suggestion, this version only prints the failure
message if the file exists or if it has an unreadable or unexecutable
parent directory.

drbrain commented on lib/rake/clean.rb:
```diff
@@ -31,9 +31,22 @@ def cleanup(file_name, opts={})
+
+    private
+
+    def self.file_already_gone?(file_name)
```
This should be an instance method, I think.  `module_function` will DTRT for us.

randycoulman replied:
I've never been very clear on how best to do stuff like this.

From my experiments, `private` seems to "cancel out" the effect of the earlier `module_function`, so I either need to add another `module_function` specifier after the `private`, or explicitly mark just this one method as a `module_function`.

I assume you'd like either of those better than the `self.` I've used here.  Do you have a preference?

drbrain replied:
Ah, right!

Then I would delete private and use `private_class_method :file_already_gone?`following the method

randycoulman committed:
Use `private_class_method` instead of private

randycoulman commented:
Oh, duh, right!  `private` doesn't apply to class methods.  How's that look now?
